### PR TITLE
[None][feat] Remove padding in Kimi K3 MLA module

### PR DESCRIPTION
**Pull request by pengbowang-nv**
## Dev Engineer Review

- Removes Kimi K3 MLA head padding and wrapper-managed checkpoint sharding.
- Uses the cuteDSL MLA decode kernel.
- Delegates MLA sharding and checkpoint loading to the owning linear modules.
- Updates `KimiMLARuntime` and `KimiK3MLAAttention` to require `ModelConfig`.
- Adds support and FP8 thresholds for 6, 12, 16, and 96 heads.
- Retains the mixed context/generation restriction for non-96 head counts.
- No configuration or test-list changes are reported.
- The earlier SM107 and SM107f support summary is not applicable.

## QA Engineer Review

No test changes.

## Description

K3 used to pad mla head from 96 to 128 due to kernel requirement. This PR removed it and use cuteDSL mla decode kernel.

This PR also removed padding and sharding in ckpt loading.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py [hunk 1]
```diff
@@ -266,7 +270,7 @@ def _is_supported_with_reason(
             return False, "CuTe DSL MLA FMHA only supports generation-only attention."
         # It is to disable mix-batch(context request + generation request) for now.
         # TODO: Eliminate high host overhead of cutedsl mla to enable mix-batch.
-        if meta.num_contexts != 0 or meta.num_generations <= 0:
+        if (meta.num_contexts != 0 and attn.num_heads != 96) or meta.num_generations <= 0:
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py:277]: Are you sure H96 does not regress? If K3 does not regress, should we also exclude other TP head_num like 6/12, i.e., using `attn.num_heads == 16`?

Comment (pengbowang-nv) [reply tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py:277]: H96 is not for perf perspective. Existing TRTLLM-Gen kernels may select a non-functional config, leading to crash without padding. With this PR the removal of padding is possible, and the perf gain is much larger than possible host overhead. K3 is a large model, also making it less susceptible to host overhead issue.

Comment (yuxianq) [reply tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py:277]: Are there functional trtllm-gen kernels for H6/H12? If not, we should also exclude them.

Comment (pengbowang-nv) [reply tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py:277]: No, trtllm-gen H6 and H12 will have no problem. H96 has problem because we have 2 tile sizes: 32 and 64, trtllm may select 64 which is not divisible. But for H6 and H12 they will be using 32.

Comment (yuxianq) [reply tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py:277]: Got it, thanks~

File: tensorrt_llm/_torch/cute_dsl_kernels/blackwell/attention/mla/mla_decode_fp16.py [hunk 2]
```diff
@@ -2606,7 +2606,9 @@ def softmax(
             row_max_new = tTR_rAcc.load().reduce(cute.ReductionOp.MAX,
                                                  row_max_new, 0)
 
-        elif cutlass.const_expr(arch >= Arch.sm_103 and arch <= Arch.sm_103f):
+        elif cutlass.const_expr(
+            (arch >= Arch.sm_103 and arch <= Arch.sm_103f)
+                or (arch >= Arch.sm_107 and arch <= Arch.sm_107f)):
```
[the remark this answers is not in the document]
Comment (BowenFu) [reply tensorrt_llm/_torch/cute_dsl_kernels/blackwell/attention/mla/mla_decode_fp16.py]: +1

[the remark this answers is not in the document]
Comment (pengbowang-nv) [reply tensorrt_llm/_torch/cute_dsl_kernels/blackwell/attention/mla/mla_decode_fp16.py]: Have tested and removed such branch as the symbol is missing from cutlass-dsl

**Comment by yunruis** [comment]
PR description issue

The auto-generated summary claims:

▎ "Added SM107 and SM107f support for the MLA softmax tensor-memory reduction path."

This is stale. Commit ac4f7d2 added the SM107 branch, but commit c9a8229 ("Revert 'add sm107 branch for mla'") reverted it. The final diff contains no SM107-related code at all. The description was generated before the revert and was never updated — please remove/regenerate that line before merging to avoid confusing reviewers and polluting release notes.

File: tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py [hunk 3]
```diff
@@ -218,16 +218,20 @@ def _is_perf_favorable(
         critical batch size (``_PERF_MIN_BATCH_FP8``); everything else falls
         back to the next FMHA library.
 
-        bf16/fp16 KV: only num_heads == 16 is admitted.
+        bf16/fp16 KV: admit the measured Kimi K3 DEP/TEP head counts and the
+        existing 16-head path.
 
         ``batch_size=None`` evaluates only the batch-size-independent
         conditions (dtype, num_heads, seq_len_q) and skips the batch floor."""
+        _BF16_FP16_PERF_FAVORABLE_HEAD_COUNTS = (6, 12, 16, 96)
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py:226]: The FP8 path was not updated to match. With an FP8 KV cache — which `modeling_kimi_linear.py` explicitly forwards to the MLA backend via `kv_cache_quant_algo` — `_get_kernel_dtype` returns `float8_e4m3fn`, and `_PERF_MIN_BATCH_FP8` below has no entries for 6, 12, or 96 heads. So every K3 decode with FP8 KV rejects CuTe DSL and falls back to trtllm-gen; at attention-DP/tp=1 that is the 96-head configuration the comment at line 272 says trtllm-gen may fail on. Before this PR, padding made this the covered `(128, 1)`/`(128, 2)` case. Either add measured FP8 entries for the new head counts, or document why FP8-KV K3 cannot reach this path. More broadly: once the fallback kernel can fail for a head count, this "perf-only" gate becomes correctness-critical — worth a comment saying so.

File: tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py [hunk 4]
```diff
@@ -265,8 +269,9 @@ def _is_supported_with_reason(
         if fwd.attention_input_type != AttentionInputType.generation_only:
             return False, "CuTe DSL MLA FMHA only supports generation-only attention."
         # It is to disable mix-batch(context request + generation request) for now.
+        # Trtllm-gen may fail under num_head == 96 so it is exempted from this check.
         # TODO: Eliminate high host overhead of cutedsl mla to enable mix-batch.
-        if meta.num_contexts != 0 or meta.num_generations <= 0:
+        if (meta.num_contexts != 0 and attn.num_heads != 96) or meta.num_generations <= 0:
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py:277]: Two issues with this exemption. (1) It is not a guarantee: a 96-head mixed batch admitted here can still be rejected by the later checks (custom/tree spec-decode masks, the FP8 batch floor, `can_implement`), and then falls through to trtllm-gen at 96 heads — the exact failure this line exists to avoid, now surfacing further from the cause. If trtllm-gen genuinely cannot run 96-head generation, consider failing loudly on that fallback (or verify those rejection paths are unreachable for K3). (2) `attn.num_heads` is the per-rank head count, so any MLA model that happens to land on 96 heads per rank now sends mixed batches through CuTe DSL and pays the host-overhead cost the TODO above describes. The comment should at least state the per-rank semantics; scoping the exemption to the attention modules it was measured on would be safer than a global head-count match.

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py [hunk 5]
```diff
@@ -2340,11 +2244,23 @@ def load_param(name: str, param: torch.nn.Parameter):
...
             if mla_mixer is not None:
-                _load_kimi_k3_mla_kv_b_proj(
-                    mla_mixer,
-                    src,
-                    head_start=mla_tp_rank * mla_mixer.num_heads,
+                h = mla_mixer.num_heads_tp
+                n = mla_mixer.qk_nope_head_dim
+                v = mla_mixer.v_head_dim
+                kv = mla_mixer.kv_lora_rank
+                local = mla_mixer.kv_b_proj.load_shard(src, device=param.device).view(h, n + v, kv)
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_kimi_linear.py:2610]: The deleted `_load_kimi_k3_mla_kv_b_proj` validated the checkpoint shape (`shape[1] == kv`, rows divisible by `n+v`) and raised descriptive errors; a malformed KV-B tensor now surfaces as an opaque `view` failure. Since this branch already computes h/n/v/kv, a one-line shape check with the old error message would keep loader failures diagnosable. Also note `h = num_heads_tp` while `v_b_proj` is allocated with `num_heads_tp_cp` (mla.py:532) — equivalent today with cp unsupported, but using `num_heads_tp_cp` here would keep the two in sync.

**Review by brnguyen2 — APPROVED** [review]
Approving — the comments below are optional touch-ups, not blockers.

The loader refactor is the right direction — deleting the hand-rolled pad-then-shard slicing in favor of the owning modules' `load_shard` removes ~190 lines of duplicated TP policy and the zero-head waste. Remaining concerns, beyond the inline comments:

- **TP coverage.** The bf16 admission list `{6, 12, 16, 96}` covers attention-DP (96) and tp=8/16 (12/6). tp=2 and tp=4 now produce 48/24 per-rank heads that fall back to trtllm-gen — are those head counts known-good on the trtllm-gen generation kernels? Also, TP sizes that don't divide 96 (e.g. 64) now assert at MLA init (`num_heads % tp == 0`), where the old 128-head padding made them work. If those configs are intentionally dropped, say so in the description.
- **No tests.** The Test Coverage section is empty and there are no test changes for a checkpoint-loader rewrite plus new kernel-admission head counts. Please state which configs were validated end-to-end (attention-DP, tp=8, tp=16; bf16 and FP8 KV cache), and consider a CPU-only unit test that loads a fake KV-B checkpoint at tp>1 and checks `kv_b_proj.weight`/`k_b_proj_trans`/`v_b_proj` against the grouped layout — that would lock in the trickiest part of this change.
- **Ticket.** This is a nontrivial feature/perf change; the title carries `[None]` — please attach the JIRA ticket.
- **Description.** The mixed-batch exemption for 96 heads is a behavior change worth a sentence in the human-written description, not just the auto-generated summary.

File: tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py (hunk 3 above)
Comment (YihuiLu512) [reply tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py:226]: +1

**Review by YihuiLu512 — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py (hunk 5 above)
Comment (pengbowang-nv) [reply tensorrt_llm/_torch/models/modeling_kimi_linear.py:2610]: h should remain num_heads_tp: it describes the TP-local layout returned by kv_b_proj.load_shard(). CP sharding, if supported, should be applied separately to v_weight using cp_rank, not by changing this reshape dimension.

File: tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py (hunk 4 above)
Comment (pengbowang-nv) [reply tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py:277]: Fp8 will be fixed. This applies to all num_heads 96, not only kimi k3.

File: tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py (hunk 3 above)
Comment (pengbowang-nv) [reply tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py:226]: 6, 12 won't have functionality issue, so only here for perf meaning. FP8 is fixed.

Comment (brnguyen2) [reply tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py:226]: Confirmed the FP8 fix (ce10387 adds the 6/12/96-head entries, with `(96, 1): 1`), so FP8-KV 96-head decode now stays on CuTe DSL instead of falling to trtllm-gen. The only residual was FP8 + MTP at 96 heads (`seq_len_q > 1`), and we don't have MTP support yet, so that path isn't reachable today. Broader fallback hardening tracked in TRTLLM-15634. Resolving.

File: tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py (hunk 4 above)
Comment (brnguyen2) [reply tensorrt_llm/_torch/attention_backend/fmha/cute_dsl_mla.py:277]: Thanks. The residual concern is that admitting a 96-head request here isn't a guarantee it stays on CuTe DSL: the custom/tree-mask check, the FP8 floor for `seq_len_q > 1`, or `can_implement` can still reject it and fall through to trtllm-gen at 96 heads. Not blocking this PR since those paths aren't exercised today. Tracked in TRTLLM-15634 (fail loudly or prove the rejection paths unreachable, and note the per-rank `attn.num_heads` semantics in the comment). Resolving.

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py (hunk 5 above)
Comment (brnguyen2) [reply tensorrt_llm/_torch/models/modeling_kimi_linear.py:2610]: Agreed on `h = num_heads_tp`, thanks for the explanation. The remaining item is the lost shape validation: a malformed KV-B tensor now surfaces as an opaque `view()` error instead of the old descriptive check. Not blocking. Tracked in TRTLLM-15634 to restore a one-line shape guard matching the fused gate/up path. Resolving.
